### ISSUE TITLE
Reconsider a better API to replace direct `Homebrew::Livecheck::Strategy::*` calls within casks

### ISSUE THREAD

**Issue by Bo98** [Homebrew/brew#16194]
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

5 casks in Homebrew/cask contain direct calls to `Homebrew::Livecheck::Strategy::*` methods: https://github.com/search?q=repo%3AHomebrew%2Fhomebrew-cask+CaskLoader&type=code

It's not a lot of casks, but I don't think these are great because:
* It is [private API](https://rubydoc.brew.sh/Homebrew/Livecheck/Strategy/ExtractPlist.html), which we avoid in these repos
* The API to call them is horrific, requiring `CaskLoader.load` to load the cask itself again.

### What is the motivation for the feature?

Seeing `CaskLoader.load("winzip")` inside `winzip` itself, which is problematic and causes issues.

### How will the feature be relevant to at least 90% of Homebrew users?

It realistically won't given its scoped to livecheck only which is maintainer-only, but an improved API will have better maintainability and avoids [bugs](https://github.com/Homebrew/homebrew-cask/pull/159614#discussion_r1385833221) associated by easy misuse, particularly when things change in `brew` without the knowledge these exist. A livecheck not reloading formulae and casks is an example of one such assumption that `brew` makes, unless you do so in a very specific way (i.e. using `__FILE__`).

### What alternatives to the feature have been considered?

Close my eyes and pretend it doesn't exist.

**Comment by bevanjkay** [comment]
@Bo98 There are [18 casks](https://github.com/search?q=repo%3AHomebrew%2Fhomebrew-cask+Homebrew%3A%3ALivecheck%3A%3AStrategy&type=code) that called `Homebrew::Livecheck::Strategy::*` for a variety of reasons. Is this a similar concern to recalling `CaskLoader.load(*)`? If so further improvement should be considered here.

**Comment by Bo98** [comment]
> @Bo98 There are [18 casks](https://github.com/search?q=repo%3AHomebrew%2Fhomebrew-cask+Homebrew%3A%3ALivecheck%3A%3AStrategy&type=code)

Ah whoops, missed some are not in submodules. Corrected the search query now.

**Comment by Bo98** [comment]
> At present, these are used in scenarios where the livecheck API needs to be re-used for a subsequent pass at usually loading page-content from a url that needs to be dynamically generated from a prior regex search.

So I guess the API improvement that could be done here is a way to specify a series of strategies rather than being restricted to just the one strategy.

**Comment by bevanjkay** [comment]
@Bo98 Is it worth updating some of the offending casks to at least use `CaskLoader.load(__FILE__)` in the interim?

**Comment by Bo98** [comment]
Probably yeah. It's necessary for them to at least work properly without `HOMEBREW_NO_INSTALL_FROM_API`.

**Comment by Bo98** [comment]
> So I guess the API improvement that could be done here is a way to specify a series of strategies rather than being restricted to just the one strategy.

An idea of what that could look like:

```rb
  livecheck do
    url "https://www.winzip.com/en/download/"
    regex(/href=.*?winzipmacedition[._-]?v?(\d+)\.dmg/i)
    strategy :page_match do |page, regex|
      major_version = page[regex, 1]
      next if major_version.blank?

      "https://download.winzip.com/winzipmacedition#{major_version}.dmg"
    end.then_strategy :extract_plist
    # or alternatively:
    end
    strategy :extract_plist 
  end
```

```rb
  livecheck do
    url "https://chromiumdash.appspot.com/fetch_releases?channel=Stable&platform=Mac"
    regex(/(\d+\.\d+\.\d+\.\d+)/i)
    strategy :json do |json, regex|
      # Find the v8 commit hash for the newest Chromium release version
      v8_hash = json.max_by { |item| Version.new(item["version"]) }.dig("hashes", "v8")
      next if v8_hash.blank?

      "https://chromium.googlesource.com/v8/v8.git/+/#{v8_hash}"
    end.then_strategy :page_match do |page, regex|
      page.scan(regex).map(&:first)
    end
  end
```

**Comment by apainintheneck** [comment]
In terms of how the DSL should look, would it be too confusing to just have multiple strategy blocks and then just evaluate them in the order they were declared. That would probably be the simplest option.

```rb
  livecheck do
    url "https://chromiumdash.appspot.com/fetch_releases?channel=Stable&platform=Mac"
    regex(/(\d+\.\d+\.\d+\.\d+)/i)
    strategy :json do |json|
      # Find the v8 commit hash for the newest Chromium release version
      v8_hash = json.max_by { |item| Version.new(item["version"]) }.dig("hashes", "v8")
      next if v8_hash.blank?

      "https://chromium.googlesource.com/v8/v8.git/+/#{v8_hash}"
    end
    strategy :page_match do |page, regex|
      page.scan(regex).map(&:first)
    end
  end
```

**Comment by MikeMcQuaid** [comment]
> In terms of how the DSL should look, would it be too confusing to just have multiple strategy blocks and then just evaluate them in the order they were declared. That would probably be the simplest option.

This is a nice idea.

---

I'm fine with either DSL but agreed that this should happen sooner rather than later and is desirable, thanks @Bo98!

**Comment by apainintheneck** [comment]
Yeah, my idea doesn't really make much sense since it doesn't address passing of information from the first strategy to the second one which is needed here.

Are we sure that this only requires a new DSL or would it also mean changing how livecheck works internally as well? It now seems to me like something that's not worth the effort. Maybe just add a linter check for the `CaskLoader.load(__FILE__)` thing and revisit this if this pattern starts getting used in more casks in the future?

**Comment by Bo98** [comment]
> Yeah, my idea doesn't really make much sense since it doesn't address passing of information from the first strategy to the second one which is needed here.

Not sure I understand the problem you're describing really. Each strategy is based on a URL of some form, no? So it would just take the previous one instead of the original.

```rb
url = (initial url, e.g. stable.url)
return nil if strategies.empty?

strategies.each do |strategy|
  url = run_strategy(url) # internally does its comparisons etc then calls DSL block
end
version = url # last url is version
```

Seem doable, though I guess it could be argued to look "weird" or something.

> It might be possible to have a nested strategy method though which could do the trick. It would essentially work like strategy.find_versions(*).fetch(:match) and we could call it strategy_matches.

That's fine too if sufficiently documented.

**Comment by samford** [comment]
Coming to this a little late, so pardon the information dump.

I went through `Homebrew::Livecheck::Strategy` references in `strategy` blocks and, after refactoring a few, we're currently at 21 uses of `#page_content`, 3 uses of `#page_header`, and the 4 `ExtractPlist#find_versions` instances.

Use of `Strategy#page_content` and `Strategy#page_headers` in `strategy` blocks is simply for making an additional HTTP request, so we're not using an another strategy in 24 out of 28. `#page_content` and `#page_headers` aren't likely to notably change in the future (if they do, additional parameters will be optional), so it would be easier to just make those public.

I have some stashed work to allow us to call `#page_content` and `#page_header` in `livecheck` blocks without the preceding module name (`Homebrew::Livecheck::Strategy`). The basic idea is to delegate references to `page_content` and `page_headers` in the `livecheck` block DSL to `Homebrew::Livecheck::Strategy`. **If we made those references public, would that help address that part of this issue?** If so, I can create a PR for that work.

-----

I wasn't previously aware of the four casks with `strategy` blocks calling `ExtractPlist#find_versions` but I agree that's not great. Three of those four casks follow a straightforward pattern of fetching a page, extracting information from that page, and using it to generate the URL to use with `ExtractPlist#find_versions`.

The other cask (`ibabel`) fetches a page, extracts information from that page, and then uses it both in the archive URL _and_ the version string that the `strategy` block returns. In this scenario, it would be necessary to pass multiple variables to the next `strategy` block (the url, and some date strings). As @apainintheneck noticed above, simply passing a `url` string won't be sufficient.

It's technically possible to support successive `strategy` blocks if we used a hash as the return type. Mapping the values to `#find_versions` args is straightforward but passing additional values to the next `strategy` block is more challenging. `strategy` blocks expect specific arguments (mostly implicitly), so they would need to be modified to handle variable arguments in a more explicit manner (more like what `Sparkle` does). Namely, the second `strategy` block argument is an optional regex, so being able to pass another argument when a regex isn't passed would require a different approach.

Even if it's technically feasible, the mental model of multiple `strategy` blocks isn't the easiest to understand, so I'm not sure that particular approach would be worth the notable implementation hurdles.

-----

A public method like the `#strategy_matches` example above that can be used to call a strategy's `#find_versions` method would be easy to implement and simple to work with in `#strategy` blocks by comparison. As in the example, `ExtractPlist` could be special-cased to avoid having to pass in a `cask` argument, which would resolve that issue.

If that's an agreeable solution, I'm happy to work on an implementation.

**Comment by apainintheneck** [comment]
`CaskLoader#load` is only needed because we don't have access to the cask in the strategy block scope, right? I wonder if we could evaluate that block in a module or class scope that has that information already.

**Comment by samford** [comment]
> `CaskLoader#load` is only needed because we don't have access to the cask in the strategy block scope, right? I wonder if we could evaluate that block in a module or class scope that has that information already.

That's correct. At the moment, the issue is that `ExtractPlist#find_versions` is called in a `strategy` block, which doesn't have access to the `Cask` object that `ExtractPlist` needs.

`strategy` blocks are typically executed in a strategy's `#versions_from_*` method. In this case, [`ExtractPlist#versions_from_items`](https://github.com/Homebrew/brew/blob/942e77d3a8498d08a07af98c84ce6bee4e9d07b4/Library/Homebrew/livecheck/strategy/extract_plist.rb#L68).

The simple solution is to pass `cask` from `#find_versions` into `#versions_from_items` and rework `ExtractPlist` `strategy` block argument parsing to be able to handle an optional `cask` argument. Thinking about it some more, that may be our only option (i.e., something like `#strategy_matches` would be called in a `strategy` block, so it also wouldn't have access to the `cask` object unless it's provided as an argument).

**Comment by carlocab** [comment]
> `CaskLoader#load` is only needed because we don't have access to the cask in the strategy block scope, right? I wonder if we could evaluate that block in a module or class scope that has that information already.

This seems to make a lot more sense than trying to reload the cask inside blocks that live inside a cask (which I think is a smell).

I've personally found it very surprising the few times I discovered (then forgot, and then re-discovered) that livecheck blocks (or, at least, blocks inside livecheck blocks) have a scope that's different from the formula.

**Comment by Bo98** [comment]
The only reason we need to access the cask within the cask is because of the current API. With a better public API like `strategy_matches` then all the current use cases disappear.

If another use case comes up, then it makes sense to maybe change what `self` points to.

**Comment by apainintheneck** [comment]
For service blocks we already make the formula available to allow us to get path information from it (https://github.com/Homebrew/brew/pull/13705). In this case, we don't even need to expose it as a method but just have it exist in the same scope so that we can use it in the internal `strategy_matches` method.

**Comment by MikeMcQuaid** [comment]
> For service blocks we already make the formula available to allow us to get path information from it (#13705). In this case, we don't even need to expose it as a method but just have it exist in the same scope so that we can use it in the internal `strategy_matches` method.

This seems like a good model to emulate here 👍🏻
